### DISCUSSION TITLE
Synology

### DISCUSSION THREAD

**Discussion by M3rofio** [Chia-Network/chia-blockchain#4558] [Q&A]
Hi i have two synology nas 
1 - 112 tb
2 - 55 tb
I don't know how to do it in docker if anyone can help

**Comment by ghost** [comment]
Hi

I am running my full node on a Synology DS916+, using the out of the box Docker. I wish I had that much space available myself. ;-)

I see two ways:
1. [Enable SSH service](https://mariushosting.com/how-to-ssh-into-a-synology-nas/), login with an admin user and [follow the official instructions](https://github.com/Chia-Network/chia-docker).
2. Use one of the forked [chia-docker images available on DockerHub](https://hub.docker.com/search?q=chia-docker&type=image). DockerHub images are available through the User Interface.

I first started with option 1, but then I ended up forking because I wanted a more comfortable way to plot.
I've forked the [official chia-docker repository](https://github.com/Chia-Network/chia-docker) into my [gldecurtins/chia-docker repo](https://github.com/gldecurtins/chia-docker).
My repository is available on DockerHub, so you'll find it in your Synology Docker Registry:
1. Log in to Synology DSM
2. Install Docker through Package Center
3. Open Docker, go to Registry, search for "gldecurtins/chia-docker", load the image. In the meantime there are other chia-docker images available, I didn't try them.
4. Read my [Readme](https://github.com/gldecurtins/chia-docker/blob/main/Readme.md), there are some differences to the official implementation
5. Launch the Container, Create an extra Terminal to figure out if everything works as expected

### Differences
- Loads the latest chia-blockchain on startup
- Different parameters
- Includes a script which could be used to trigger plotting on the NAS

### Challenges
- Set the log output to stdout, so that you can read them through Docker
- It took some time to sync the blockchain and wallet, make sure you have the /root/.chia folder properly mapped
- Find a good way to retrieve finished plots. I ended up enabling ftp on the NAS, using Scheduled Tasks to move finished uploads into the final folder.

Happy farming!
-Luca.

![Screenshot 2021-05-11 at 18 41 59](https://user-images.githubusercontent.com/35364122/117853448-b7919580-b288-11eb-91f6-3726beba116b.png)

**Reply by M3rofio** [reply]
can you provide me with your email sir 

i have 2 nas
synology  DS3617xs  INTEL Xeon D-1527 16gb ram - 112 tb 
synology DS3612xs   INTEL Core i3-2100 6gb ram - 55tb

**Reply by M3rofio** [reply]
hello @gldecurtins 
how to transfer my wallet and how to get the farmer public key and pool key

**Reply by bluechia** [reply]
Hello Luca, thank you for this post. After reading it it was pretty simple to make it work. I have just one additional question, can you please explain in your readme file what is the difference between start values. In my case I'd like to plot outside of Synology and use the docker on Synology only for farming. What is exactly difference between farmer and farmer-only modes? Thank you

**Reply by ghost** [reply]
Hi @bluechia, I got these start values from the chia command:
```
root@chia:/opt/chia-blockchain# venv/bin/chia start --help
Usage: chia start [OPTIONS] [all|node|harvester|farmer|farmer-no-
                  wallet|farmer-only|timelord|timelord-only|timelord-launcher-
                  only|wallet|wallet-only|introducer|simulator]...

Options:
  -r, --restart  Restart running services
  -h, --help     Show this message and exit.
```

If you execute *chia start farmer*, then it will actually start a full_node, including wallet, farmer and harvester.

On my Synology I start the farmer, which doesn't trigger the plotting process. To start the plotting process I've mapped the required volumes and setup a scheduled task with the following command:
```
docker exec -d chia /usr/local/bin/plots_create.sh /mnt/chia-plots/tmp /mnt/chia-plots/final
```
If you do not want to plot on the Synology, then just don't execute the script.

**Comment by M3rofio** [comment]
dear @gldecurtins I failed to do it can you help me with my email: m3rofio@gmail.com, or in dicord : m3rofio#0127 please contact me

**Comment by JaroslavHorky** [comment]
Hello, I'm an amateur when it comes to Docker. I have it installed and running gldecurtins/chia-docker. Would anyone advise me step by step how to further set up a Chia node?

Thank you

**Reply by ghost** [reply]
Hi @JaroslavHorky 

I would use the official docker image: https://github.com/Chia-Network/chia-docker

I've created the gldecurtins/chia-docker one to be executed on a Synology NAS without going through the ssh console, plus I wanted to learn about Docker and refresh my shell scripting skills. :-)

**Reply by JaroslavHorky** [reply]
Hi @gldecurtins, 

I also create the same and executed on my Synology. But I do not have sheel scripting skills and I do not know how to continue.

I understand that I will executed run the installed image. Then I'll go to the Terminal tab. And here I enter the following commands to run the Ful node for Chia farming?

```
docker run --name <container-name> \
    --volume /path/to/.chia:/root/.chia --volume /path/to/plots:/mnt/chia-plots/final \
    --env keys="/root/.chia/mnemonic.txt" \
    -d gldecurtins/chia-docker:latest
```
Is there anything else you need to do? I don't want to create plots on Synology.

Thank you for advice

**Reply by ghost** [reply]
Hello again

Once you've loaded and configured the docker container, then it should actually already run the farmer. Have a look at my configuration.

![Screenshot 2021-05-19 at 11 40 49](https://user-images.githubusercontent.com/35364122/118791686-3beea480-b897-11eb-994a-67a367145144.png)

The Terminal tab you can use to create a new bash shell. Inside the bash shell I usually execute some commands to validate if the farmer is working properly.
For example:
```
/opt/chia-blockchain/venv/bin/chia show -s
/opt/chia-blockchain/venv/bin/chia show -c
/opt/chia-blockchain/venv/bin/chia farm summary
/opt/chia-blockchain/venv/bin/chia wallet show
```

But first, try to get the container up and running, check the *Log* tab. Then log in and execute */opt/chia-blockchain/venv/bin/chia --help*. Also check if your volumes are properly mounted.

**Reply by ghost** [reply]
One challenge is also to setup the keys properly.. and store the words in the mnemonic.txt.
So another important command is
```
/opt/chia-blockchain/venv/bin/chia keys --help
```

If you do not have keys, then you have to generate yourself a pair and write down the words. **Do not share** these words with anyone.

**Reply by JaroslavHorky** [reply]
Hi @gldecurtins again. 

Thanks to your advice, I'm starting to orientate myself in this.

![image](https://user-images.githubusercontent.com/60477700/118806586-909a1b80-b8a7-11eb-8027-bda85c5c246b.png)

I do not currently have a defined Volumes and I do not have keys specified. 
I have a Chia final folder on NAS. I just don't know how to connect it here.
Do I have to move the plot files somewhere or am I just mapping the path? Or how do I map it to a specific folder? Will I do everything via the terminal or in the File station?

How are the keys filled? I currently have a wallet on Windows. Is it possible to move it somehow? Or do I have to add all the words over the terminal?
If via terminal, how is it written?

Somehow like this?
docker exec -it <container-name> venv/bin/chia keys add **word1**
docker exec -it <container-name> venv/bin/chia keys add **word2**
...

![image](https://user-images.githubusercontent.com/60477700/118807986-42861780-b8a9-11eb-8ac3-42796a98383a.png)

Last questions: why do I have 2x Full node there?

![image](https://user-images.githubusercontent.com/60477700/118808017-4ade5280-b8a9-11eb-959f-974e89f3586c.png)

Thank you

**Reply by ghost** [reply]
Hello

You need to stop your container in order to edit the environment variables and volume mappings. This can only be done if the container is not already running.
So stop the container, then add the volume mappings (Volume), check if you have TCP port 8444 forwarded (Port Settings) and validate the environment variables.

![Screenshot 2021-05-19 at 13 55 26](https://user-images.githubusercontent.com/35364122/118808756-1d45d900-b8aa-11eb-8c1f-6f6e686f9a8d.png)

If you keep the environment variable "keys" set to "generate", then the container will generate new keys upon start.
So, for example:
1. save your words from the windows machine into mnemonic.txt
2. Put the file into a folder which is shared with the container. e.g. docker/chia
3. Validate that docker/chia is mapped to /root/.chia
4. Set the keys environment variable to the path to the file inside of the container. e.g. /root/.chia/mnemonic.txt

**Reply by JaroslavHorky** [reply]
Hi @gldecurtins 

Thank you for further instructions. I moved a little further.

I have two directories mapped:

![image](https://user-images.githubusercontent.com/60477700/118826453-ccd77700-b8bb-11eb-8fb6-04ece6842b7b.png)

![image](https://user-images.githubusercontent.com/60477700/118826776-0f994f00-b8bc-11eb-8cc8-9cec4ce0efb3.png)

I think I have the right ports.

![image](https://user-images.githubusercontent.com/60477700/118830219-0e1d5600-b8bf-11eb-9646-8ecb67912516.png)

Now the mnemonic.txt How do I create it? Is it a plain text file or does it have a special structure? Words separated by a space, or 4 words per line, ...

Thanks

**Reply by ghost** [reply]
Hi

Ports are not yet correct. Set the Local Port to 8444, and 8555. You can delete the 58444 if you do not connect to the testnet.

Then generate some keys, which you copy into the text file:
```
/opt/chia-blockchain/venv/bin/chia keys generate
Generating private key
Added private key with public key fingerprint 1993128631 and mnemonic
room total wear toe crazy soft bla guilt roast about tomorrow media reveal inspire ramp cherry indoor figure purpose raise mansion donor disorder test
```
This will print out a fingerprint and 24 words. Take these 24 words and put them into **1 line** in your text file.

I assume that you've generated a few keys in the meantime. So execute:
```
/opt/chia-blockchain/venv/bin/chia keys show
```
Then you can delete the keys you no longer want with:
```
/opt/chia-blockchain/venv/bin/chia keys delete -f <Fingerprint>
```

**Reply by JaroslavHorky** [reply]
so like this?:

![image](https://user-images.githubusercontent.com/60477700/118833721-f72c3300-b8c1-11eb-853d-0c303e0c6624.png)

So if I already have the keys from Windows, then I don't have to generate anything and just put them in a file and upload them to a folder?

Somehow my terminal stopped working. I can write there, but nothing is done. What could it be?

![image](https://user-images.githubusercontent.com/60477700/118837200-e0d3a680-b8c4-11eb-876d-5c25cb356e33.png)


Thank you for your patience with me

**Reply by ghost** [reply]
Hi again

No, like this:
![Screenshot 2021-05-19 at 18 56 43](https://user-images.githubusercontent.com/35364122/118853585-33688f00-b8d4-11eb-984f-dcfbe7589042.png)

Also take care that port 8444 is forwarded from your Router to your NAS.

Then open a new terminal, press the "Create" button:
![Screenshot 2021-05-19 at 18 58 01](https://user-images.githubusercontent.com/35364122/118853678-4b401300-b8d4-11eb-8c4e-7fcc29b1e013.png)

Inside the second terminal you can execute all kind of commands, inside the container. You can Delete it again, or start multiple... 

You can take your windows words, put the into the text file, set the keys value and start the container. On startup there should be a message stating that it has used your key, including fingerprint and words.

**Comment by MEBruck** [comment]
Thank you Fldecurtins for posting this info and your answers to JaroslavHorky answered a lot of my questions (also a newbie to Docker). However, I do have another question, if you don't mind...
I have a full node running on Windows and just wanted to setup my Synology NAS for farming. I was able to get your container setup and running (it recognizes my existing plots, which happen to be stored on the same NAS). However, I'm unsure what to do to link this Docker install of Chia with my original main node. I want that main node to stop farming and have that fully taken over by the the Chia running on my NAS. I don't really see the next steps in the instructions to finish that part up. Any help you can provide would be GREATLY appreciated. I hate to say it, I'm also a Windows-guy NOT a Unix person, so any extra detail on your instructions/steps to help me out there is also appreciated.

**Reply by ghost** [reply]
Hi MEBruck

I think you need to figure out a strategy on how to move the finished plots from your windows machine to your NAS.
The Docker image points to a shared Synology folder e.g. Chia/plots. The Chia software scans the plots directory/directories every 2 minutes. If new plots were added in the meantime, then these will be included to the farm automatically.

So:

1. Move all your finished plots to the NAS
2. Configure the docker image so that it points to the NAS folder with the plots
3. Configure the firewall to forward TCP 8444 to the NAS instead of the Windows mache
4. Start the Docker container, pointing to the folder with the plots

Then find a way to move finished plots from the Windows machine to the NAS. I use a script to copy files over using ftp, but you can also use Network shares.

Regards and hope this helps

**Reply by happygreens** [reply]
hi,

 I have 3 chia wallets with 200 plots on nas, already faming on virtual windows 10machine with one payout-address. 

May I use your scenario for this in docker?nning Your system is already ru
Is there an easy way to transfer it? The pathes are aready set.... I think I just need to set some input lines...

Parameter to pay with:
wallet_1
wallet_2
wallet_3
payoutadress_wallet1

**Reply by happygreens** [reply]
I got wallet 2 and 3 later... therefore I changed payout adress to wallet 1. This already farmes... But I like to have it on your system. I would appreciate your help.
